### PR TITLE
[GlobalIsel] Combine G_SELECT

### PR DESCRIPTION
**Pull request by tschuett**
Cleanups, preparation for more combines, add known bits for constant conditions, combine selects where the false and true register are constants, and improve support for vector conditions.

AMDGPU supports vector conditions. X86 has a todo for vector conditions. AArch64 SVE supports SEL for vector conditions. How to implement vector conditions with NEON (with bsl), see arm64-vselect.ll ? Vector select asserts in the instruction selector.

buildNot does not support scalable vectors. We cannot create scalable constant vectors of -1 and there is no G_Not. AArch64 SVE has a NOT and a DUP for broadcasting. Something akin to G_CONSTANT_SPLAT, G_CONSTANT_VECTOR, G_SPLAT_VECTOR, G_BRODCAST, or G_HOMOGENOUS_VECTOR that takes an immediate and creates a (fixed or scalable) vector where all elements are the immediate might solve the buildNot challenge, facilitates new combines, pattern matching, and new selecting optimizations.

P.S. We need to support integer and float.

https://github.com/llvm/llvm-project/pull/74502

```c
<vscale x 4 x i32> splat (i32 -1)
```

### CODE REVIEW

File: llvm/test/CodeGen/AArch64/GlobalISel/combine-select.mir [hunk 1]
```diff
@@ -253,9 +253,7 @@ body:             |
     ; CHECK-NEXT: [[COPY1:%[0-9]+]]:_(s64) = COPY $x1
     ; CHECK-NEXT: %c:_(s1) = G_TRUNC [[COPY]](s64)
     ; CHECK-NEXT: %t:_(s1) = G_TRUNC [[COPY1]](s64)
-    ; CHECK-NEXT: %one:_(s1) = G_CONSTANT i1 true
-    ; CHECK-NEXT: [[XOR:%[0-9]+]]:_(s1) = G_XOR %c, %one
-    ; CHECK-NEXT: %sel:_(s1) = G_OR [[XOR]], %t
+    ; CHECK-NEXT: %sel:_(s1) = G_OR %c, %t
```
[the remark this answers is not in the document]
Comment (tschuett) [reply llvm/test/CodeGen/AArch64/GlobalISel/combine-select.mir:256]: The DAGCombiner has a function: foldBoolSelectToLogic. 
`select Cond, T, 1 --> or (not Cond), T`
The not (XOR -1) seems to be missing.

File: llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp [hunk 2]
```diff
@@ -6318,3 +6246,343 @@ void CombinerHelper::applyCommuteBinOpOperands(MachineInstr &MI) {
...
+    return true;
+  }
+
+  // select Cond, T, 1 --> or (not Cond), T
+  if (isOneOrOneSplat(False, /* AllowUndefs */ true)) {
+    MatchInfo = [=](MachineIRBuilder &B) {
+      B.setInstrAndDebugLoc(*Select);
+      // First the not.
+      Register Inner = MRI.createGenericVirtualRegister(CondTy);
+      B.buildNot(Inner, Cond);
```
Comment (tschuett) [inline llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp:6508]: The`Ìnner` Is dead.

File: llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp [hunk 3]
```diff
@@ -6318,3 +6246,343 @@ void CombinerHelper::applyCommuteBinOpOperands(MachineInstr &MI) {
...
+      // Then an ext to match the destination register.
+      Register Ext = MRI.createGenericVirtualRegister(TrueTy);
+      B.buildZExtOrTrunc(Ext, Cond);
```
Comment (tschuett) [inline llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp:6511]: This should ext the inner and not the cond.

Comment (arsenm) [reply llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp:6511]: Hide the MRI.createGenericVirtualRegister in the MIRBuilder call

Comment (tschuett) [reply llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp:6511]: How? I need these registers.

**Comment by tschuett** [comment]
This check will fail once scalable vector splats get merged:

https://github.com/llvm/llvm-project/blob/bfebadc8c392fc3bc8cbd54af735a006f50a5d8c/llvm/lib/CodeGen/GlobalISel/MachineIRBuilder.cpp#L317

**Pull request closed without merging**